### PR TITLE
[xmppclient] New XMPP (Jabber) messaging binding (deprecated -> see #5268)

### PR DESCRIPTION
**Pull request by pavel-gololobov**
New XMPP (Jabber) binding

This is a new version of XMPP (Jabber) binding, which is based on the newest version of the Smack Library. Binding allows you to send and receive messages via xmpp using new Actions interface.

For example:

```
rule "Leak detected"
when
    Item Xi_Leak changed
then
    if(Xi_Leak.state == ON) {
        val actions = getActions("xmpp","xmppclient:xmppBridge:xmpp")
        actions.publishXMPP("pavel@example.com","Warning! Leak detected!")
    }
end
```

And process received messages like this:

```
rule "Turn off all lights with separator and reply"
when
    Channel "xmppclient:xmppBridge:xmpp:xmpp_command" triggered
then
    var actionName = receivedEvent.getEvent().split("#")
    if(actionName.get(1).toLowerCase() == "turn off lights") {
        Group_Light_Home_All.sendCommand(OFF)

        val actions = getActions("xmpp","xmppclient:xmppBridge:xmpp")
        actions.publishXMPP(actionName.get(0),"All lights was turned off")
    }
end
``` 

Signed-off-by: Pavel Gololobov <pavel@ochat.org> (github: pavel-gololobov)

### CODE REVIEW

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/action/XMPPActions.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    public static void publishXMPP(@Nullable ThingActions actions, @Nullable String to, @Nullable String text) {
+        if(actions == null) {
+            logger.error("Sending error, actions is NULL");
+            throw new IllegalArgumentException("actions is NULL");
+        }
+        if (actions instanceof XMPPActions) {
+            ((XMPPActions) actions).publishXMPP(to, text);
+        } else {
+            logger.error("Sending error, instance of actions is {}", actions.getClass().getName());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/action/XMPPActions.java]: Please don't use the error log level that is reserved to the framework.

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * The {@link XMPPClientConfiguration} class contains fields mapping thing configuration parameters.
+ *
+ * @author Pavel Gololobov - Initial contribution
+ */
+@NonNullByDefault
+public class XMPPClientConfiguration {
+
+    public String host;
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientConfiguration.java]: Does that compile? `NonNullByDefault` usually forces you to assign default values or annotate your fields with `@Nullable`.

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientHandler.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        return xmppClient;
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(XMPPActions.class);
+    }
+
+    /**
+     * Does nothing
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientHandler.java]: Please move: Less is better sometimes

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientMessageSubscriber.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * Subscriber interface
+ * 
+ * @author Pavel Gololobov - Initial contribution
+ */
+public interface XMPPClientMessageSubscriber {
+
+    void processMessage(String from, String payload);
+    String getName();
+    
+}
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientMessageSubscriber.java]: newline missing

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClient.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            reconnectionManager.enableAutomaticReconnection();
+
+            ReconnectionManager.setEnabledPerDefault(true);
+
+            connection.connect().login();
+
+            ChatManager chatManager = ChatManager.getInstanceFor(connection);
+            chatManager.addIncomingListener(this);
+        } catch (SmackException | IOException | XMPPException | InterruptedException e) {
+            logger.error("XMPP CONNECTION ERROR", e);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClient.java]: logger

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClient.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        try {
+            ChatManager chatManager = ChatManager.getInstanceFor(connection);
+            chatManager.addIncomingListener(this);
+
+            EntityBareJid jid = JidCreate.entityBareFrom(to);
+            Chat chat = chatManager.chatWith(jid);
+            chat.send(message);
+        } catch (XmppStringprepException | SmackException.NotConnectedException | InterruptedException e) {
+            logger.error("XMPP SENDING ERROR", e);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClient.java]: logger. And in your message: Don't shout at your innocence user. You can't blame him for this to have happened.

File: addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientMessageSubscriber.java (hunk 4 above)
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientMessageSubscriber.java]: I mean at the end of the file. In github you see this little icon for files without newlines. All files are required to have a newline except the manifest file (which ceases to exist on the new buildsystem though)

Comment (pavel-gololobov) [reply addons/binding/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/handler/XMPPClientMessageSubscriber.java]: Ok

**Comment by pavel-gololobov** [comment]
I removed the binding from the old location and migrated to the bnd build system. Binding successfully build. Only I did not understand how to manage external libraries. JAR file is built without dependencies.

**Comment by davidgraeff** [comment]
You forgot to remove the old files.
And yes the dependencies are not part of the generated jar. They are loaded at runtime. That's why you need to define them in the karaf feature file. Have a look in that file for other bundles using that mechanism (search for `dependency="true"`).

**Comment by davidgraeff** [comment]
Can you please remove the `bundles/org.openhab.binding.xmppclient/.factorypath` file? I don't know which eclipse plugin is generating that one, but it seems like it should be part of a `.gitignore` file as it is probably auto generated.

Also remove the .settings directory. If you rebase your openhab2-addons repo you will find a newer .gitignore that prevents those files from being added to a commit.

Other that that, I think we are ready to merge, as soon as the conflict is gone and travis succeeded.

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 7]
```diff
@@ -859,6 +859,15 @@
         <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.windcentrale/${project.version}</bundle>
     </feature>
 
+    <feature name="openhab-binding-xmppclient" description="XMPP Client Binding" version="${project.version}">
+        <feature>openhab-runtime-base</feature>
+        <bundle dependency="true" start-level="80">mvn:org.igniterealtime.smack/smack-java7/4.3.1</bundle>
```
Comment (davidgraeff) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: The compilation itself worked on travis but the feature verification failed. Please remove the start-level attributes from the dependencies. And add "wrap" to the URI like in "wrap:mvn:org.igniterealtime.smack/smack-java7/4.3.1" (and for the other deps as well)

Comment (pavel-gololobov) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Done. But I don’t understand how the DCO check works...

**Comment by davidgraeff** [comment]
You would need to squash all your commits (git rebase -i HEAD~20, pick your commits except the first one and use 's' as the first word) and sign them: "-s" on the command line.

**Comment by clinique** [comment]
Should this binding replace the 1.x existing action ?

**Comment by pavel-gololobov** [comment]
Sorry, I am not very familiar with github. My primary VCS is Mercurial for long time. I no longer understand what happened to my branch. If everything is bad, I can try to make the new PR, and you remove this one. Excuse me.

**Comment by davidgraeff** [comment]
Yup, you can also create a new one. You seem to have merged changes from a different branch to yours. Always do a "git pull -r origin master" (for rebasing on the master). In Guis it is probably called "pull with rebase" or similar.

**Comment by pavel-gololobov** [comment]
David, thank you. I created a new PR.: https://github.com/openhab/openhab2-addons/pull/5268

**Pull request closed without merging**